### PR TITLE
ST: New approach for kafka client usage

### PR DESCRIPTION
**Pull request by Frawless**
### Type of change

- Bugfix
- Enhancement / new feature
- Refactoring

### Description

Since Job used for kafka clients wasn't 100% stable I decide to deploy clients as a deployment. Test client image now contains API for http requests. These requests are controling the clients, which can be started/stopped from system tests. Pod with test clients can handle multiple clients (python, java, c++) in the future and can run clients in parallel.

### Checklist

- [x] Write tests
- [x] Make sure all tests pass
- [x] Update documentation
- [x] Update CHANGELOG.md

### CODE REVIEW

**Comment by Frawless** [comment]
@strimzi-ci run tests tags=all

**Comment by Frawless** [comment]
@strimzi-ci run tests tags=all

**Comment by Frawless** [comment]
@strimzi-ci run tests tags=""

**Comment by Frawless** [comment]
@strimzi-ci run tests testcase=KafkaST tags=acceptance,regression,flaky,cci_flaky

**Comment by Frawless** [comment]
@strimzi-ci tags=acceptance

**Comment by Frawless** [comment]
@strimzi-ci tags=acceptance,regression

**Comment by Frawless** [comment]
@strimzi-ci tags=acceptance,regression

**Comment by Frawless** [comment]
@strimzi-ci tags=acceptance,regression

**Comment by Frawless** [comment]
@strimzi-ci run tests tags=acceptance,regression

**Comment by Frawless** [comment]
@strimzi-ci run tests tags=acceptance,regression

**Comment by Frawless** [comment]
@strimzi-ci run tests tags=acceptance,regression

**Comment by Frawless** [comment]
@strimzi-ci run tests tags=acceptance,regression

File: systemtest/src/main/java/io/strimzi/systemtest/Environment.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+package io.strimzi.systemtest;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class Environment {
+    private static final Logger LOGGER = LogManager.getLogger(Environment.class);
+    private static Environment instance;
+
+    public static final String STRIMZI_ORG_ENV = "STRIMZI_ORG";
```
[the remark this answers is not in the document]
Comment (Frawless) [reply systemtest/src/main/java/io/strimzi/systemtest/Environment.java]: Why? Now it's not used widely but I plan to do that. Don't see any benefit with make it private now and public in another PR

File: systemtest/src/main/java/io/strimzi/systemtest/Environment.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    }
+
+    public static synchronized Environment getInstance() {
+        if (instance == null) {
+            instance = new Environment();
+        }
+        return instance;
+    }
+
+    public String getStrimziOrg() {
```
[the remark this answers is not in the document]
Comment (Frawless) [reply systemtest/src/main/java/io/strimzi/systemtest/Environment.java]: Good point, but it should be used. I probably didn't change all occurrences for these variables. I want to have only one class for all constants/env variables for ST.

File: systemtest/src/main/java/io/strimzi/systemtest/MessagingBaseST.java [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+        cliApiClient = new MsgCliApiClient(new URL("http://" + KAFKA_CLIENTS + "." + clientUrl + ":80"));
+    }
+
+    /**
+     * Simple availability check for kafka cluster
+     * @param producer producer
+     * @param consumer consumer
+     * @param clusterName cluster name
+     */
+    void availabilityTest(AbstractClient producer, AbstractClient consumer, String clusterName) throws InterruptedException, ExecutionException, TimeoutException {
```
[the remark this answers is not in the document]
Comment (Frawless) [reply systemtest/src/main/java/io/strimzi/systemtest/MessagingBaseST.java]: Is that a problem? Lets say it's some kind of our send-receive msg API. Definitely someone of us will use it

File: systemtest/src/main/java/io/strimzi/systemtest/MessagingBaseST.java [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+     * @param producer producer
+     * @param consumer consumer
+     * @param messageCount message count
+     * @param timeout timeout for producer and consumer to be finished
+     * @param clusterName cluster name
+     * @param tlsListener option for tls listener inside kafka cluster
+     * @param topicName topic name
+     * @param user user for tls if it's used for messages
+     */
+    void availabilityTest(AbstractClient producer, AbstractClient consumer, int messageCount, int timeout, String clusterName, boolean tlsListener, String topicName, KafkaUser user) throws InterruptedException, ExecutionException, TimeoutException {
```
[the remark this answers is not in the document]
Comment (Frawless) [reply systemtest/src/main/java/io/strimzi/systemtest/MessagingBaseST.java]: Any idea? We want't all of them in that case. It's overridden few times where some parameters are constants, guess it's enough.

File: systemtest/src/main/java/io/strimzi/systemtest/MessagingBaseST.java [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+     * @param producer producer
+     * @param messageCount messages count
+     * @param timeout timeout for producer to be finished
+     * @param clusterName cluster name
+     * @param tlsListener option for tls listener inside kafka cluster
+     * @param topicName topic name
+     * @param user user for tls if it's used for messages
+     * @return count of send and acknowledged messages
+     */
+    int sendMessages(AbstractClient producer, int messageCount, int timeout, String clusterName, boolean tlsListener, String topicName, KafkaUser user) throws InterruptedException, ExecutionException, TimeoutException {
```
[the remark this answers is not in the document]
Comment (Frawless) [reply systemtest/src/main/java/io/strimzi/systemtest/MessagingBaseST.java]: Client's arguments are different. We can create one method, but it will be hard to read the code I think. Now from method name you know what exactly it's doing. We can cut out some things from these methods, but we should keep both of them. For call only one method we have `availabilityTest` method which combines functionality of send and receive.

File: systemtest/src/main/java/io/strimzi/systemtest/MessagingBaseST.java [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+     * @param timeout timeout
+     */
+    private void waitTillProcessFinish(String processUuid, String description, int timeout) {
+        TestUtils.waitFor("Wait till " + description + " finished", 2000, timeout, () -> {
+            JsonObject out;
+            try {
+                out = cliApiClient.getClientInfo(processUuid);
+                setResponse(out);
+                return !out.getBoolean("isRunning");
+            } catch (InterruptedException | ExecutionException | TimeoutException e) {
```
[the remark this answers is not in the document]
Comment (Frawless) [reply systemtest/src/main/java/io/strimzi/systemtest/MessagingBaseST.java]: Put there some log message is maybe better for getting info from test log.

File: systemtest/src/main/java/io/strimzi/systemtest/apiclients/MsgCliApiClient.java [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+    /**
+     * Stop messaging webClient and get all informations about them (stdOut, stdErr, code, isRunning)
+     *
+     * @param uuid webClient id
+     * @return
+     * @throws InterruptedException
+     * @throws ExecutionException
+     * @throws TimeoutException
+     */
+    public JsonObject stopClient(String uuid) throws InterruptedException, ExecutionException, TimeoutException {
```
[the remark this answers is not in the document]
Comment (Frawless) [reply systemtest/src/main/java/io/strimzi/systemtest/apiclients/MsgCliApiClient.java]: API approach. In the future we can have a case where we will need it

File: systemtest/src/main/java/io/strimzi/systemtest/apiclients/MsgCliApiClient.java [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+
+        JsonObject response = startClients(apiArgument, 1);
+        JsonArray ids = response.getJsonArray("clients");
+        String uuid = ids.getString(0);
+
+        Thread.sleep(5000);
+
+        response = getClientInfo(uuid);
+        response.put("UUID", uuid);
+        return response;
```
[the remark this answers is not in the document]
Comment (Frawless) [reply systemtest/src/main/java/io/strimzi/systemtest/apiclients/MsgCliApiClient.java]: You can investigate it, but I don't think it's connected to fabric8 somehow

File: systemtest/src/main/java/io/strimzi/systemtest/apiclients/MsgCliApiClient.java [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+    }
+
+    /***
+     * Send request with one webClient and receive id
+     * @param client
+     * @return id of webClient
+     */
+    public String sendAndGetId(AbstractClient client) throws Exception {
```
[the remark this answers is not in the document]
Comment (Frawless) [reply systemtest/src/main/java/io/strimzi/systemtest/apiclients/MsgCliApiClient.java]: Same as with `stopClient()`.

File: systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/ClientArgument.java [hunk 10]
```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright 2018, Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.strimzi.systemtest.kafkaclients;
+
+/**
+ * Enum with argument for external clients
+ */
+public enum ClientArgument {
```
[the remark this answers is not in the document]
Comment (Frawless) [reply systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/ClientArgument.java]: Ye we don;t actually. It was designed for console consumer as well, but I decided to remove it.

**Comment by Frawless** [comment]
@strimzi-ci run tests testcase=KafkaST#testSendMessagesPlainAnonymous+testSendMessagesTlsAuthenticated+testSendMessagesPlainScramSha+testSendMessagesTlsScramSha

**Comment by Frawless** [comment]
@strimzi-ci run tests testcase=KafkaST#testSendMessagesPlainAnonymous+testSendMessagesTlsAuthenticated+testSendMessagesPlainScramSha+testSendMessagesTlsScramSha

**Comment by Frawless** [comment]
@strimzi-ci run tests testcase=KafkaST#testSendMessagesPlainAnonymous+testSendMessagesTlsAuthenticated+testSendMessagesPlainScramSha+testSendMessagesTlsScramSha

**Comment by Frawless** [comment]
@strimzi-ci run tests testcase=KafkaST#testSendMessagesPlainAnonymous+testSendMessagesTlsAuthenticated+testSendMessagesPlainScramSha+testSendMessagesTlsScramSha tags=acceptance,regression

**Comment by Frawless** [comment]
@strimzi-ci run tests testcase=KafkaST#testSendMessagesPlainAnonymous+testSendMessagesTlsAuthenticated+testSendMessagesPlainScramSha+testSendMessagesTlsScramSha tags=acceptance,regression

File: test/src/main/java/io/strimzi/test/k8s/Exec.java [hunk 11]
```diff
@@ -9,111 +9,285 @@
...
-                    .redirectOutput(out)
-                    .redirectError(err);
-            Process p = pb.start();
-            OutputStream outputStream = p.getOutputStream();
-            if (input != null) {
-                LOGGER.trace("With stdin {}", input);
-                outputStream.write(input.getBytes(Charset.defaultCharset()));
+            Exec executor = new Exec();
+            ret = executor.execute(input, command, timeout);
+            synchronized (LOCK) {
```
Comment (tombentley) [inline test/src/main/java/io/strimzi/test/k8s/Exec.java]: Why is this necessary?

File: test/src/main/java/io/strimzi/test/k8s/Exec.java [hunk 12]
```diff
@@ -9,111 +9,285 @@
...
 import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
 import java.nio.file.Files;
-import java.util.ArrayList;
+import java.nio.file.Path;
+import java.nio.file.Paths;
 import java.util.Arrays;
 import java.util.List;
+import java.util.Scanner;
+import java.util.concurrent.CompletableFuture;
```
Comment (tombentley) [inline test/src/main/java/io/strimzi/test/k8s/Exec.java]: The rest of the project is using vertx `Futures`. It's not clear to me whether whatever you're doing with the `CompletableFuture` API here cannot be done using vertx, but if possible we should avoid using two different async APIs in the same project.

File: test/src/main/java/io/strimzi/test/k8s/KubeClient.java [hunk 13]
```diff
@@ -93,7 +93,7 @@ default K delete(String... files) {
      * @param command The command
      * @return The process result.
      */
-    ProcessResult execInPod(String pod, String... command);
+    ExecResult execInPod(String pod, String... command);
```
Comment (tombentley) [inline test/src/main/java/io/strimzi/test/k8s/KubeClient.java]: Why rename the class?

File: systemtest/src/main/java/io/strimzi/systemtest/apiclients/MsgCliApiClient.java [hunk 14]
```diff
@@ -0,0 +1,233 @@
...
+     */
+    public JsonObject startClients(List<String> clientArguments, int count) throws Exception {
+        Future<JsonObject> responsePromise = Future.future();
+//        CompletableFuture<JsonObject> responsePromise = new CompletableFuture<>();
+        JsonObject request = new JsonObject();
+        request.put("command", new JsonArray(clientArguments));
+        request.put("count", count);
+
+        return doRequestNTimes(initRetry, () -> {
+                webClient.post(endpoint.getPort(), endpoint.getHost(), "")
```
Comment (tombentley) [inline systemtest/src/main/java/io/strimzi/systemtest/apiclients/MsgCliApiClient.java]: Is the request URL really the empty string? (I would have expected `"/"` as least)

File: systemtest/src/main/java/io/strimzi/systemtest/apiclients/MsgCliApiClient.java [hunk 15]
```diff
@@ -0,0 +1,233 @@
...
+     *
+     * @param clientArguments list of arguments for kafka client (together with kafka client name!)
+     * @param count           count of clients that will be started
+     * @return
+     * @throws InterruptedException
+     * @throws ExecutionException
+     * @throws TimeoutException
+     */
+    public JsonObject startClients(List<String> clientArguments, int count) throws Exception {
+        Future<JsonObject> responsePromise = Future.future();
```
Comment (tombentley) [inline systemtest/src/main/java/io/strimzi/systemtest/apiclients/MsgCliApiClient.java]: You need to move this line inside the lambda you're passing to `doRequestNTimes()`, because otherwise the recursive calls to `doRequestNTimes()` will end up trying to complete the future _again_ (which is not allowed. `Futures` are one-shot).

File: systemtest/src/main/java/io/strimzi/systemtest/apiclients/MsgCliApiClient.java [hunk 16]
```diff
@@ -0,0 +1,233 @@
...
+                        .sendJson(request,
+                                ar -> responseHandler(ar, responsePromise, HttpURLConnection.HTTP_OK, "Error starting messaging clients"));
+
+            // Here I should wait till responsePromies will be completed somehow
+            // After putting there breakpoint, responsePromise will never be completed (evn without bp)
+            if (responsePromise.result() == null) {
+                throw new Exception("Result is null!");
+            }
+
+            return responsePromise.result();
```
Comment (tombentley) [inline systemtest/src/main/java/io/strimzi/systemtest/apiclients/MsgCliApiClient.java]: You're returning here before you know whether the `responseHandler()` has been called. The lambda argument to `sendJson()` will be called asynchronously . If you want a blocking lambda to pass to `doRequestNTimes()` then you need to do the blocking yourself. Vertx doesn't provide any nice way to say "block until this `Future` is completed", so you might not want to use the vertx web client at all, but rather some other HTTP client which has the blocking API you seem to want to be using.

File: .findbugs/findbugs-exclude.xml [hunk 17]
```diff
@@ -16,5 +16,9 @@
   <Match>
     <Class name="~io\.fabric8\.kubernetes\.api\.model\.(Doneable).+(\$.*)?" />
   </Match>
+  <Match>
+    <Bug pattern="DM_DEFAULT_ENCODING"/>
+    <Class name="~io\.strimzi\.test\.k8s\.Exec.*" />
```
[the remark this answers is not in the document]
Comment (Frawless) [reply .findbugs/findbugs-exclude.xml]: I had problem there with encoding even when I set default utf-8. Let me try to remove it and see what's going.

File: test/src/main/java/io/strimzi/test/k8s/KubeClient.java (hunk 13 above)
Comment (Frawless) [reply test/src/main/java/io/strimzi/test/k8s/KubeClient.java]: It make more sense to me - we have `Exec` as an executor so results of this should be `ExecResult`

File: test/src/main/java/io/strimzi/test/k8s/Exec.java (hunk 11 above)
Comment (Frawless) [reply test/src/main/java/io/strimzi/test/k8s/Exec.java]: It's for keep some order in logging in case we have multiple executor running on the same time

**Comment by Frawless** [comment]
@strimzi-ci run tests testcase=SecurityST,KafkaST tags=acceptance,regression,cci_flaky,flaky

File: pom.xml [hunk 18]
```diff
@@ -302,6 +303,12 @@
                 <artifactId>vertx-core</artifactId>
                 <version>${vertx.version}</version>
             </dependency>
+<!--            <dependency>-->
```
Comment (serrss) [inline pom.xml]: Should you remove this code?

File: test/src/main/java/io/strimzi/test/TestUtils.java [hunk 19]
```diff
@@ -433,5 +437,40 @@ public static String getContent(File file, Function<JsonNode, String> edit) {
...
     }
+
+    /**
+     * Repeat request n-times in a row
+     *
+     * @param retry count of remaining retries
+     * @param fn    request function
+     * @return
+     */
+    public static <T> T doRequestNTimes(int retry, Callable<T> fn, Optional<Runnable> reconnect) throws Exception {
```
Comment (serrss) [inline test/src/main/java/io/strimzi/test/TestUtils.java]: Why do we need to send request 10 times in MsgCliApiClient? Sending one request is not enough?

File: systemtest/src/main/java/io/strimzi/systemtest/Resources.java [hunk 20]
```diff
@@ -741,6 +827,241 @@ DoneableKubernetesClusterRoleBinding kubernetesClusterRoleBinding(KubernetesClus
...
+
+    DoneableIngress createIngress(String appName, int port, String url, String clientNamespace) throws Exception {
+        Ingress ingress = getSystemtestIngressResource(appName, port, url);
+        LOGGER.info("Creating ingress {} in namespace {}", ingress.getMetadata().getName(), clientNamespace);
+        client.extensions().ingresses().inNamespace(clientNamespace).create(ingress);
+        deleteLater(ingress);
+        return new DoneableIngress(ingress);
+    }
+
+    private PodSpec createClientSpec(boolean tlsListener, KafkaUser... kafkaUsers) {
```
Comment (serrss) [inline systemtest/src/main/java/io/strimzi/systemtest/Resources.java]: Can we move this logic to a different location? `Resources` is a really big class =(

File: test/src/main/java/io/strimzi/test/TestUtils.java (hunk 19 above)
Comment (Frawless) [reply test/src/main/java/io/strimzi/test/TestUtils.java]: The html request sometimes fails, this should workaround this. It send one request and based on the result it will sent another or end.

File: systemtest/src/main/java/io/strimzi/systemtest/Resources.java (hunk 20 above)
Comment (Frawless) [reply systemtest/src/main/java/io/strimzi/systemtest/Resources.java]: You mean only this method? Guess it's possible to do that, but I don't see not enough benefit from it. Maybe it will be better to move out all methods from `Resources` which could be static and it doesn't matter from where they are called. You can create issue for that and we can manage it in some next PR.

**Comment by Frawless** [comment]
@strimzi-ci run tests tags=regression testcase=testSendMessagesPlainAnonymous+testSendMessagesTlsAuthenticated+testSendMessagesPlainScramSha+testSendMessagesTlsScramSha

**Comment by Frawless** [comment]
@strimzi-ci run tests tags=regression testcase=KafkaST#testSendMessagesPlainAnonymous+testSendMessagesTlsAuthenticated+testSendMessagesPlainScramSha+testSendMessagesTlsScramSha

**Comment by Frawless** [comment]
@strimzi-ci run tests tags=regression testcase=KafkaST#testSendMessagesPlainAnonymous+testSendMessagesTlsAuthenticated+testSendMessagesPlainScramSha+testSendMessagesTlsScramSha

**Comment by Frawless** [comment]
@strimzi-ci run tests testcase=KafkaST#testSendMessagesPlainAnonymous+testSendMessagesTlsAuthenticated+testSendMessagesPlainScramSha+testSendMessagesTlsScramSha tags=acceptance,regression
